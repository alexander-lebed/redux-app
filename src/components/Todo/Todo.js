// @flow
import React from 'react';
import { Map } from 'immutable';
import Modal from 'react-bootstrap/lib/Modal';
import ButtonToolbar from 'react-bootstrap/lib/ButtonToolbar';
import Button from 'react-bootstrap/lib/Button';
import Glyphicon from 'react-bootstrap/lib/Glyphicon';


type Props = {
    todo: Map<string, any>,
    onEdit: (id: string, text: string) => void,
    onDelete: (id: string) => void
}

type State = {
    todoToEdit: Map<string, any> | null
}

export default class Todo extends React.Component<void, Props, State> {
    state: State = {
        todoToEdit: null
    };

    onEdit = (event: Object) => {
        if (event) {
            // stop the click event from bubbling up
            event.stopPropagation();
        }
        this.setState({
            todoToEdit: this.props.todo
        });
    }

    editTodo = (todo: Map<string, any>) => {
        this.props.onEdit(todo.get('id'), todo.get('text'));
        this.setState({ todoToEdit: null });
    }

    onSubmit = (event: Object) => {
        const isEnterKey = event.which === 13;
        if (isEnterKey && this.state.todoToEdit) {
            this.editTodo(this.state.todoToEdit);
        }
    }

    render() {
        const { todo, onDelete } = this.props;
        const { todoToEdit } = this.state;

        const stateSetter = obj => this.setState(obj);

        let editModal;
        if (todoToEdit) {
            editModal = (
                <Modal show onHide={() => stateSetter({ todoToEdit: null })}>
                    <Modal.Header>
                        <Modal.Title>Edit todo</Modal.Title>
                    </Modal.Header>
                    <Modal.Body>
                        <input
                            type="text"
                            className='todo__entry'
                            value={todoToEdit.get('text')}
                            onChange={e => stateSetter({ todoToEdit: todoToEdit.set('text', e.target.value) })}
                            onKeyDown={this.onSubmit}
                        />
                    </Modal.Body>
                    <Modal.Footer>
                        <Button
                            id='confirm-btn'
                            className='pull-right'
                            disabled={false}
                            onClick={() => this.editTodo(todoToEdit)}
                        >
                            EDIT
                        </Button>
                    </Modal.Footer>
                </Modal>
            );
        } else {
            editModal = null;
        }

        return (
            <div>
                {todo.get('isDone') ?
                    <strike>{todo.get('text')}</strike> :
                    <span>{todo.get('text')}</span>
                }
                <ButtonToolbar className="pull-right" style={{ marginTop: 3 }}>
                    <Glyphicon
                        id="edit-todo"
                        glyph="pencil"
                        style={{ marginRight: 15, color: 'grey' }}
                        onClick={e => this.onEdit(e)}
                    />
                    <Glyphicon
                        id="remove-todo"
                        glyph="remove"
                        style={{ color: 'grey' }}
                        onClick={onDelete}
                    />
                </ButtonToolbar>
                {editModal && editModal}
            </div>
        );
    }
}
