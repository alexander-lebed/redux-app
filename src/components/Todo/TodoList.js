// @flow
/* eslint-disable no-shadow, no-param-reassign */
import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Col from 'react-bootstrap/lib/Col';
import Row from 'react-bootstrap/lib/Row';
import type { State, Dispatch } from '../../redux/types';
import { addTodo, updateTodo, toggleTodo, removeTodo } from '../../redux/actions/todo-actions';
import Todo from './Todo';


type Props = {
    currentData: State,
    addTodo: (text: string) => void,
    updateTodo: (id: string, text: string) => void,
    toggleTodo: (id: string) => void,
    removeTodo: (id: string) => void
}

export class TodoList extends React.Component<void, Props, void> {

    onSubmit = (event: Object) => {
        const text = event.target.value;
        const isEnterKey = event.which === 13;
        if (isEnterKey && text.length > 0) {
            this.props.addTodo(text);
            event.target.value = '';
        }
    }

    render() {
        const { currentData, updateTodo, toggleTodo, removeTodo } = this.props;
        const todos = currentData.get('todos');
        return (
            <div className="todo">
                <Row>
                    <Col xs={3} />
                    <Col xs={6}>
                        <input
                            type="text"
                            className='todo__entry'
                            placeholder="Add todo"
                            onKeyDown={this.onSubmit}
                        />
                        <ul className="todo__list">
                            {todos.filter(t => !t.get('isDeleted')).map(t => (
                                <li
                                    key={t.get('id')}
                                    className="todo__item"
                                    onClick={() => toggleTodo(t.get('id'))}
                                >
                                    <Todo
                                        todo={t}
                                        onEdit={(id, text) => updateTodo(id, text)}
                                        onDelete={() => removeTodo(t.get('id'))}
                                    />
                                </li>
                            ))}
                        </ul>
                    </Col>
                    <Col xs={3} />
                </Row>
            </div>
        );
    }
}

export default connect(
    // mapStateToProps
    state => ({ currentData: state.todoData }),
    // mapDispatchToProps
    (dispatch: Dispatch) =>
        bindActionCreators({ addTodo, updateTodo, toggleTodo, removeTodo }, dispatch)
        // return {
        //     addTodo: text => dispatch(addTodo(text)),
        //     toggleTodo: id => dispatch(toggleTodo(id)),
        //     removeTodo: id => dispatch(removeTodo(id))
        // };

)(TodoList);
