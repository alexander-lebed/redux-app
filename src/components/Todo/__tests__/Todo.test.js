// @flow
import React from 'react';
import Modal from 'react-bootstrap/lib/Modal';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { Map } from 'immutable';
import Todo from '../Todo';


describe('<Todo/>', () => {
    test('should render Todo', () => {
        const todo = Map({ id: 1, isDone: false, text: 'some text', isDeleted: false });
        const wrapper = renderer.create(
            <Todo todo={todo} onEdit={jest.fn()} onDelete={jest.fn()} />
        );
        const tree = wrapper.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('should render crossed out Todo as it\'s done', () => {
        const todo = Map({ id: 1, isDone: true, text: 'some text', isDeleted: false });
        const wrapper = shallow(
            <Todo todo={todo} onEdit={jest.fn()} onDelete={jest.fn()} />
        );
        expect(wrapper.find('strike').length).toEqual(1);
    });

    test('should render edit modal on pencil click', () => {
        const todo = Map({ id: 1, isDone: true, text: 'some text', isDeleted: false });
        const wrapper = shallow(
            <Todo todo={todo} onEdit={jest.fn()} onDelete={jest.fn()} />
        );
        wrapper.find('#edit-todo').simulate('click');
        expect(wrapper.find(Modal).length).toEqual(1);
    });
});

