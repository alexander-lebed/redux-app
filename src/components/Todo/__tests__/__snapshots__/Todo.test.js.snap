// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todo/> should render Todo 1`] = `
<div>
  <span>
    some text
  </span>
  <div
    className="pull-right btn-toolbar"
    role="toolbar"
    style={
      Object {
        "marginTop": 3,
      }
    }
  >
    <span
      className="glyphicon glyphicon-pencil"
      id="edit-todo"
      onClick={[Function]}
      style={
        Object {
          "color": "grey",
          "marginRight": 15,
        }
      }
    />
    <span
      className="glyphicon glyphicon-remove"
      id="remove-todo"
      onClick={[Function]}
      style={
        Object {
          "color": "grey",
        }
      }
    />
  </div>
</div>
`;
